### PR TITLE
Add ignore-unknown-fields flag

### PR DESCRIPTION
**Pull request by codesoap**
I'm happily picking up the suggestion from #143 to add an `ignoreUnknownFields` flag with this pull request. Similarly to the mempool flag, it can either be used as an option in the .proto file, or as a flag on the CLI.

I was able to test the change with a command line flag on the CLI, but was unable to test it with the option in the .proto file. Unfortunately I still don't fully understand how to use the `ext.proto` file in a project (the import `google/protobuf/descriptor.proto` failed).

I'm also unsure whether I should add new test cases for the feature and if so, how.

### CODE REVIEW

**Review by vmg — CHANGES_REQUESTED** [review]
Good start! We should definitely add some tests for individual messages. https://github.com/planetscale/vtprotobuf/blob/main/testproto/pool/pool.proto is an example of testing an ext flag; simply adding another package under `testproto` and ensuring that the generated code compiles successfully will be enough to test this feature, as all it's doing is removing fields from the codegen.

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+	if m == nil {
+		return 0
+	}
+	var l int
+	_ = l
+	l = len(m.Foo)
+	if l > 0 {
+		n += 1 + l + protohelpers.SizeOfVarint(uint64(l))
+	}
+	n += len(m.unknownFields)
```
Comment (vmg) [inline testproto/ignore_unknown_fields/opt_vtproto.pb.go:149]: This can be omitted.

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+func (m *IgnoreUnknownFieldsExtension) CloneVT() *IgnoreUnknownFieldsExtension {
+	if m == nil {
+		return (*IgnoreUnknownFieldsExtension)(nil)
+	}
+	r := new(IgnoreUnknownFieldsExtension)
+	r.Foo = m.Foo
+	if len(m.unknownFields) > 0 {
+		r.unknownFields = make([]byte, len(m.unknownFields))
+		copy(r.unknownFields, m.unknownFields)
+	}
```
Comment (vmg) [inline testproto/ignore_unknown_fields/opt_vtproto.pb.go:32]: This can be omitted.

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+		return 0, nil
+	}
+	i := len(dAtA)
+	_ = i
+	var l int
+	_ = l
+	if m.unknownFields != nil {
+		i -= len(m.unknownFields)
+		copy(dAtA[i:], m.unknownFields)
+	}
```
Comment (vmg) [inline testproto/ignore_unknown_fields/opt_vtproto.pb.go:128]: This can be omitted.

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
```
Comment (vmg) [inline testproto/ignore_unknown_fields/opt_vtproto.pb.go:88]: This can be omitted.

**Review by vmg — CHANGES_REQUESTED** [review]
Great! Now to wrap this up, I think we should make the flag more consistent and ignore unknown fields _everywhere_ (in marshalling and size calculations), not only when unmarshalling. I've commented on the leftover usages that can be removed from codegen when the flag is set.

**Comment by codesoap** [comment]
I had not changed the code you mentioned, because it wouldn't do anything, if `unknownFields` hadn't been filled during the unmarshalling. If `unknownFields` was still present, someone must have used the "regular" `Unmarshal` instead of `UnmarshalVT` or maybe had stored an old message, from whence the `ignoreUnknownFields` feature was not active.

I wasn't quite sure what the best approach in such a scenario would be and thought it might be best to assume the `unknownFields` were present for a good reason. However, I'm fine with your approach as well and am now stripping the `unknownFields` out of every message when calculating size, cloning and marshalling.

I have not extended the documentation for the newly adapted methods, as I feel like the changes won't be interesting for most of the users.

PS: I assume you are aware, but just to be sure: The `EqualVT` method still reports equality only if the `unknownFields` match.

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+func (this *IgnoreUnknownFieldsExtension) EqualVT(that *IgnoreUnknownFieldsExtension) bool {
+	if this == that {
+		return true
+	} else if this == nil || that == nil {
+		return false
+	}
+	if this.Foo != that.Foo {
+		return false
+	}
+	return string(this.unknownFields) == string(that.unknownFields)
```
Comment (vmg) [inline testproto/ignore_unknown_fields/opt_vtproto.pb.go:45]: Sorry! You missed this one! Please give a quick review to the all the codegen steps to ensure that nothing else is using `unknownFields`. A `git grep` for `*.go` files will find everything. :)

**Comment by vmg** [comment]
> PS: I assume you are aware, but just to be sure: The EqualVT method still reports equality only if the unknownFields match.

I think it makes a lot of sense to _not_ fail equality when you're working on a message where you've explicitly opted in to ignore unknown fields. Likewise, if you've Unmarshalled with protobuf, we still don't want to emit unknown fields when you serialize with this library, and so on. Any other behavior would be IMO inconsistent, so fully cleaning up the codegen of all `unknownFields` references seems sensible to me.

PS: thanks again for all your effort here!

File: testproto/ignore_unknown_fields/opt_vtproto.pb.go (hunk 5 above)
Comment (codesoap) [reply testproto/ignore_unknown_fields/opt_vtproto.pb.go]: I've changed it.

**Review by vmg — APPROVED** [review]

**Comment by vmg** [comment]
Thank you!

**Comment by codesoap** [comment]
Thanks a lot for your help and patience!

**Pull request merged**